driver/src/main/java/org/neo4j/driver/internal/async/pool/ConnectionPoolImpl.java

```diff
@@ -86,27 +93,30 @@ public ConnectionPoolImpl( ChannelConnector connector, Bootstrap bootstrap, Pool
     }
 
     @Override
-    public void purge( BoltServerAddress address )
+    public void retainAll( Set<BoltServerAddress> addressesToRetain )
     {
-        log.info( "Purging connections towards %s", address );
-
-        // purge active connections
-        activeChannelTracker.purge( address );
-
-        // purge idle connections in the pool and pool itself
-        ChannelPool pool = pools.remove( address );
-        if ( pool != null )
+        for ( BoltServerAddress address : pools.keySet() )
         {
-            pool.close();
+            if ( !addressesToRetain.contains( address ) )
+            {
+                int activeChannels = activeChannelTracker.activeChannelCount( address );
+                if ( activeChannels == 0 )
```

Comment:
Will we also purge all `IDLE` connection if the activeChannel is not 0?
If so then maybe also a test: `shouldCloseIdleConnInPoolWhenRetaining`?

FYI: 

 Driver | remove from RT  | close idle  | disallow return to pool 
 ------ | --------------- | ---------- | -------------------------------- 
 JS       | * o | * o | * o 
 python      | * o     |   * o | X X 
 .NET | * o      |    * o | * o 
 Java 1.4 | * o | * o | * o 
 Java 1.5 | * o | X X | X X 

[*]connection error
[o]not in new routing table

Reply:
No, it will not purge idle connections in that case. Whole pool will be closed when `activeChannels == 0` and address is not in the routing table.